File: crates/core/src/retrieval.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use time::OffsetDateTime;

use crate::{
    chunk::ChunkId,
    diagnosis::EvidenceDiagnosisSummary,
    embedding::{ChunkEmbedding, EmbeddingModelInfo},
    ingestion::ChunkPreview,
    model::ModelConfigId,
    source::{Document, DocumentId, Source, SourceId},
    trace::TraceId,
};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RetrievalRun {
    pub id: RetrievalRunId,
    pub trace_id: TraceId,
    pub query: String,
    pub rewritten_query: Option<String>,
    pub retriever: RetrieverKind,
    pub embedding_model_id: ModelConfigId,
    pub top_k: u32,
    pub hits: Vec<RetrievalHit>,
    pub latency_ms: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RetrievalHit {
    pub chunk_id: ChunkId,
    pub rank: u32,
    pub score: f32,
    pub citation: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub enum RetrieverKind {
    Vector,
    Keyword,
    Hybrid,
    Custom(String),
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq, Hash)]
pub struct RetrievalRunId(pub Uuid);

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub struct RetrievalQueryRequest {
    pub query: String,
    #[serde(default = "default_retrieval_top_k")]
    pub top_k: u32,
    #[serde(default)]
    pub retrieval_mode: RetrievalMode,
    #[serde(default)]
    pub source_ids: Vec<SourceId>,
    #[serde(default)]
    pub document_ids: Vec<DocumentId>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RetrievalQueryResponse {
    pub run: RetrievalQueryRun,
    pub answer: ExtractiveAnswer,
    pub hits: Vec<RetrievalQueryHit>,
    pub embedding_status: RetrievalEmbeddingStatus,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub diagnosis: Option<EvidenceDiagnosisSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RetrievalQueryRun {
    pub id: RetrievalQueryRunId,
    pub query: String,
    pub top_k: u32,
    pub retrieval_mode: RetrievalMode,
    pub latency_ms: u64,
    #[serde(with = "crate::wire_time")]
    pub created_at: OffsetDateTime,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RetrievalQueryHit {
    pub rank: u32,
    pub score: f32,
    pub chunk: ChunkPreview,
    pub document: Document,
    pub source: Source,
    pub matched_terms: Vec<RetrievalMatchedTerm>,
    pub score_breakdown: RetrievalScoreBreakdown,
    pub normalized_score_breakdown: RetrievalScoreBreakdown,
    pub snippet: String,
    pub citation: RetrievalCitation,
    #[serde(default)]
    pub quality_flags: Vec<RetrievalQualityFlag>,
    pub evidence_strength: EvidenceStrength,
    pub duplicate_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub struct RetrievalMatchedTerm {
    pub term: String,
    pub count: u32,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
pub struct RetrievalScoreBreakdown {
    pub semantic: f32,
    pub lexical: f32,
    pub phrase: f32,
    pub section: f32,
    pub path: f32,
    pub metadata: f32,
}

impl RetrievalScoreBreakdown {
    pub fn zero() -> Self {
        Self {
            semantic: 0.0,
            lexical: 0.0,
            phrase: 0.0,
            section: 0.0,
            path: 0.0,
            metadata: 0.0,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq, Hash)]
#[serde(rename_all = "snake_case")]
pub enum RetrievalQualityFlag {
    Duplicate,
    HeadingOnly,
    TooShort,
    WeakEvidence,
    SemanticMatch,
    ExactTermMatch,
    SectionOnlyMatch,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum EvidenceStrength {
    Strong,
    Medium,
    Weak,
}

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub struct RetrievalCitation {
    pub label: String,
    pub chunk_id: ChunkId,
    pub document_id: DocumentId,
    pub document_path: String,
    pub chunk_ordinal: u32,
    pub section_title: Option<String>,
    pub checksum_prefix: String,
    pub snippet: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub struct ExtractiveAnswer {
    pub status: ExtractiveAnswerStatus,
    pub text: String,
    pub citations: Vec<RetrievalCitation>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum ExtractiveAnswerStatus {
    Answered,
    InsufficientEvidence,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SearchableChunk {
    pub source: Source,
    pub document: Document,
    pub chunk: crate::chunk::Chunk,
    pub embedding: Option<ChunkEmbedding>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq, Hash)]
pub struct RetrievalQueryRunId(pub Uuid);

#[derive(Debug, Default, Clone, Copy, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RetrievalMode {
    Lexical,
    Vector,
    #[default]
    Hybrid,
}

impl RetrievalMode {
    pub fn requires_embeddings(self) -> bool {
        matches!(self, Self::Vector | Self::Hybrid)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
pub struct RetrievalEmbeddingStatus {
    pub readiness: RetrievalEmbeddingReadiness,
    pub required: bool,
    pub model: EmbeddingModelInfo,
    pub total_chunks: u32,
    pub indexed_chunks: u32,
    pub missing_chunks: u32,
    pub stale_chunks: u32,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RetrievalEmbeddingReadiness {
    NotRequired,
    Ready,
    Partial,
    Missing,
}

pub const DEFAULT_RETRIEVAL_TOP_K: u32 = 5;

fn default_retrieval_top_k() -> u32 {
    DEFAULT_RETRIEVAL_TOP_K
}

File: apps/web/src/lib/api/retrieval.ts
import { jsonRequest, requestJson } from "./client";
import type { ChunkPreview, DocumentRecord, SourceRecord } from "./sources";
import type { RetrievalEmbeddingStatus } from "./embeddings";
import type {
  CreateRetrievalEvalCaseRequest,
  RetrievalEvalCase,
  RetrievalEvalRun,
  RunRetrievalEvalRequest,
} from "./evalLab";

export type RetrievalMode = "lexical" | "vector" | "hybrid";
export type EvidenceStrength = "strong" | "medium" | "weak";
export type ExtractiveAnswerStatus = "answered" | "insufficient_evidence";
export type RetrievalQualityFlag =
  | "duplicate"
  | "heading_only"
  | "too_short"
  | "weak_evidence"
  | "semantic_match"
  | "exact_term_match"
  | "section_only_match";
export type DiagnosisOutcome = "strong" | "mixed" | "weak" | "failing";
export type DiagnosisSeverity = "info" | "warning" | "critical";
export type DiagnosisFailureCode =
  | "missing_document"
  | "missing_embedding_index"
  | "partial_embedding_index"
  | "weak_evidence"
  | "duplicate_evidence"
  | "heading_only_evidence"
  | "low_score_margin"
  | "vector_lexical_disagreement"
  | "citation_missing"
  | "top_result_not_cited"
  | "missing_expected_evidence";
export type DiagnosisRemediationArea =
  | "chunking"
  | "embeddings"
  | "top_k"
  | "retrieval_mode"
  | "reranking"
  | "metadata_filters"
  | "citations"
  | "corpus_coverage"
  | "other";
export type DiagnosisRecommendationPriority =
  | "critical"
  | "high"
  | "medium"
  | "low";
export type DiagnosisScoreSignal =
  | "semantic"
  | "lexical"
  | "phrase"
  | "section"
  | "path"
  | "metadata"
  | "none";

export interface DiagnosisFailure {
  code: DiagnosisFailureCode;
  severity: DiagnosisSeverity;
  title: string;
  summary: string;
  evidence_refs: string[];
}

export interface EvidenceScoreExplanation {
  evidence_ref: string;
  chunk_id: string;
  rank: number;
  final_score: number;
  score_delta_from_previous: number | null;
  score_delta_to_next: number | null;
  dominant_signal: DiagnosisScoreSignal;
  score_breakdown: RetrievalScoreBreakdown;
  normalized_score_breakdown: RetrievalScoreBreakdown;
  summary: string;
}

export interface DiagnosisRecommendation {
  code: string;
  priority: DiagnosisRecommendationPriority;
  area: DiagnosisRemediationArea;
  title: string;
  rationale: string;
  action: string;
  failure_codes: DiagnosisFailureCode[];
  evidence_refs: string[];
}

export interface EvidenceDiagnosisSummary {
  outcome: DiagnosisOutcome;
  summary: string;
  primary_issue: DiagnosisFailure | null;
  failures: DiagnosisFailure[];
  score_explanations: EvidenceScoreExplanation[];
  recommendations: DiagnosisRecommendation[];
}

export interface RetrievalQueryRequest {
  query: string;
  top_k?: number;
  retrieval_mode?: RetrievalMode;
  source_ids?: string[];
  document_ids?: string[];
}

export interface RetrievalQueryRun {
  id: string;
  query: string;
  top_k: number;
  retrieval_mode: RetrievalMode;
  latency_ms: number;
  created_at: string;
}

export interface RetrievalMatchedTerm {
  term: string;
  count: number;
}

export interface RetrievalScoreBreakdown {
  semantic: number;
  lexical: number;
  phrase: number;
  section: number;
  path: number;
  metadata: number;
}

export interface RetrievalCitation {
  label: string;
  chunk_id: string;
  document_id: string;
  document_path: string;
  chunk_ordinal: number;
  section_title: string | null;
  checksum_prefix: string;
  snippet: string;
}

export interface ExtractiveAnswer {
  status: ExtractiveAnswerStatus;
  text: string;
  citations: RetrievalCitation[];
}

export interface RetrievalQueryHit {
  rank: number;
  score: number;
  chunk: ChunkPreview;
  document: DocumentRecord;
  source: SourceRecord;
  matched_terms: RetrievalMatchedTerm[];
  score_breakdown: RetrievalScoreBreakdown;
  normalized_score_breakdown: RetrievalScoreBreakdown;
  snippet: string;
  citation: RetrievalCitation;
  quality_flags: RetrievalQualityFlag[];
  evidence_strength: EvidenceStrength;
  duplicate_count: number;
}

export interface RetrievalQueryResponse {
  run: RetrievalQueryRun;
  answer: ExtractiveAnswer;
  hits: RetrievalQueryHit[];
  embedding_status: RetrievalEmbeddingStatus;
  diagnosis?: EvidenceDiagnosisSummary | null;
}

export function queryRetrieval(
  request: RetrievalQueryRequest,
  signal?: AbortSignal,
): Promise<RetrievalQueryResponse> {
  return requestJson<RetrievalQueryResponse>(
    "/api/v1/retrieval/query",
    jsonRequest("POST", request, signal),
  );
}

export function listRetrievalEvalCases(
  signal?: AbortSignal,
): Promise<RetrievalEvalCase[]> {
  return requestJson<RetrievalEvalCase[]>("/api/v1/retrieval/evals", {
    signal,
  });
}

export function createRetrievalEvalCase(
  request: CreateRetrievalEvalCaseRequest,
  signal?: AbortSignal,
): Promise<RetrievalEvalCase> {
  return requestJson<RetrievalEvalCase>(
    "/api/v1/retrieval/evals",
    jsonRequest("POST", request, signal),
  );
}

export function runRetrievalEvals(
  request: RunRetrievalEvalRequest,
  signal?: AbortSignal,
): Promise<RetrievalEvalRun> {
  return requestJson<RetrievalEvalRun>(
    "/api/v1/retrieval/evals/run",
    jsonRequest("POST", request, signal),
  );
}
